src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -0,0 +1,314 @@
+
+            return IsValidMSBuildPath(msBuildExePath)
+                ? new BuildEnvironment(msBuildExePath, runningTests, runningInVisualStudio, visualStudioPath)
+                : null;
+        }
+
+        private static BuildEnvironment TryFromFolder(string folder, bool runningTests, bool runningInVisualStudio, string visualStudioPath)
+        {
+            if (string.IsNullOrEmpty(folder)) return null;
+
+            var msBuildPath = Path.Combine(folder, "MSBuild.exe");
+
+            return IsValidMSBuildPath(msBuildPath)
+                ? new BuildEnvironment(msBuildPath, runningTests, runningInVisualStudio, visualStudioPath)
+                : null;
+        }
+
+        /// <summary>
+        /// Determine whether the given path is considered to be an acceptable path to MSBuild.
+        /// </summary>
+        /// <remarks>
+        /// If we are running in an orphaned way (i.e. running from Microsoft.Build.dll in someone else's process),
+        /// that folder will not be sufficient as a build tools folder (e.g. we can't launch MSBuild.exe from that
+        /// location). At minimum, it must have MSBuild.exe and MSBuild.exe.config.
+        /// </remarks>
+        /// <param name="path">Full path to MSBuild.exe</param>
+        /// <returns>True when the path to MSBuild is valid.</returns>
+        private static bool IsValidMSBuildPath(string path)
+        {
+            return !string.IsNullOrEmpty(path) &&
+                   Path.GetFileName(path).Equals("MSBuild.exe", StringComparison.OrdinalIgnoreCase) &&
+                   File.Exists(path) &&
+                   File.Exists($"{path}.config");
+        }
+
+        /// <summary>
+        /// Look for Visual Studio
+        /// </summary>
+        /// <param name="processName"></param>
+        /// <param name="processName2"></param>
+        /// <param name="runningInVisualStudio"></param>
+        /// <returns></returns>
+        private static string FindVisualStudio(string[] processNames, out bool runningInVisualStudio)
+        {
+            runningInVisualStudio = false;
+
+            // Check to see if we're running inside of Visual Studio
+            foreach (var process in processNames)
+            {
+                if (IsProcessInList(process, s_visualStudioProcess))
+                {
+                    runningInVisualStudio = true;
+
+                    // This assumes running from VS\Common7\IDE\devenv.exe.
+                    return FileUtilities.GetFolderAbove(process, 3);
+                }
+            }
+
+            // VSInstallDir is set from the Developer Command Prompt
+            var vsInstallDir = Environment.GetEnvironmentVariable("VSINSTALLDIR");
+            var vsVersion = Environment.GetEnvironmentVariable("VisualStudioVersion");
+
+            if (!string.IsNullOrEmpty(vsInstallDir) &&
+                !string.IsNullOrEmpty(vsVersion) &&
+                vsVersion == "15.0" &&
```

Comment:
Is this something that's going to have to be updated with each new VS version? If so, can you replace it with something like (Float.Parse(vsVersion) >= 15)?